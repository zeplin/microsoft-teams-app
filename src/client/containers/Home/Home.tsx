import React, {
    FunctionComponent, Reducer, useEffect, useReducer
} from "react";
import * as microsoftTeams from "@microsoft/teams-js";
import { Loader } from "@fluentui/react-northstar";
import { Login } from "./components";

enum Status {
    LOADING,
    LOGIN,
    CONFIGURATION
}

enum ActionType {
    COMPLETE_LOADING,
    GET_TOKEN
}

type State = {
    status: Status.LOADING | Status.LOGIN;
} | {
    status: Status.CONFIGURATION;
    accessToken: string;
}

type Action = {
    type: ActionType.COMPLETE_LOADING;
} | {
    type: ActionType.GET_TOKEN;
    value: string;
}

const reducer: Reducer<State, Action> = (state, action) => {
    switch (action.type) {
        case ActionType.COMPLETE_LOADING:
            return { status: Status.LOGIN };
        case ActionType.GET_TOKEN:
            return { status: Status.CONFIGURATION, accessToken: action.value };
        default:
            throw new Error();
    }
};

export const Home: FunctionComponent = () => {
    const [state, dispatch] = useReducer(reducer, { status: Status.LOADING });

    useEffect(() => {
        microsoftTeams.initialize(() => {
            microsoftTeams.appInitialization.notifySuccess();
            dispatch({ type: ActionType.COMPLETE_LOADING });
        });
    }, []);

    switch (state.status) {
        case Status.LOADING:
            return <Loader id="loader-home"/>;
        case Status.LOGIN:
            return <Login onTokenReceive={(value): void => dispatch({ type: ActionType.GET_TOKEN, value })} />;
        case Status.CONFIGURATION:
            return (
                <div>
                    {state.accessToken}
                </div>
            );
        default:
            throw new Error();
    }
};
