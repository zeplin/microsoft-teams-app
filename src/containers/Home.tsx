import React, {
    FunctionComponent, useEffect, useRef, useState
} from "react";
import * as microsoftTeams from "@microsoft/teams-js";
import { Provider, Flex, Input } from "@fluentui/react-northstar";
import { BASE_URL } from "config";

const Home: FunctionComponent = () => {
    const [input, setInput] = useState("");
    const [isLoading, setIsLoading] = useState(true);

    /**
     * Workaround: microsoftTeams.settings.registerOnSaveHandler have to be called only once.
     * When the variable `input` is dependency of the useEffect of register functions, then the sdk produce error
     * For more info: https://reactjs.org/docs/hooks-faq.html#how-to-read-an-often-changing-value-from-usecallback
     */
    const inputRef = useRef(input);

    useEffect(() => {
        inputRef.current = input;
    }, [input]);

    useEffect(() => {
        microsoftTeams.initialize(() => {
            microsoftTeams.appInitialization.notifySuccess();
            microsoftTeams.settings.getSettings(({ entityId }) => {
                setInput(entityId);
                setIsLoading(false);
            });
        });

        microsoftTeams.settings.registerOnSaveHandler(event => {
            const settings = {
                entityId: inputRef.current,
                contentUrl: `${BASE_URL}/?name={loginHint}&tenant={tid}&group={groupId}&theme={theme}`,
                configName: inputRef.current
            };
            microsoftTeams.settings.setSettings(settings);
            event.notifySuccess();
        });

        microsoftTeams.settings.registerOnRemoveHandler(event => {
            event.notifySuccess();
        });
    }, []);

    return (
        <Provider>
            <Flex fill={true}>
                <Flex.Item>
                    <Input
                        disabled={isLoading}
                        value={input}
                        onChange={(ignore, { value }): void => {
                            setInput(value);
                            microsoftTeams.settings.setValidityState(Boolean(value));
                        }}
                    />
                </Flex.Item>
            </Flex>
        </Provider>
    );
};

export default Home;
