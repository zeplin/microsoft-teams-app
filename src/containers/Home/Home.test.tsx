import React from "react";
import { Home } from "./Home";
import * as microsoftTeams from "@microsoft/teams-js";
import { fireEvent, render } from "@testing-library/react";

jest.mock("@microsoft/teams-js", () => ({
    initialize: jest.fn(callback => callback()),
    appInitialization: {
        notifySuccess: jest.fn()
    },
    settings: {
        registerOnSaveHandler: jest.fn(),
        registerOnRemoveHandler: jest.fn(),
        setValidityState: jest.fn(),
        setSettings: jest.fn(),
        getSettings: jest.fn(callback => callback({ entityId: "" }))
    }
}));

jest.mock("../../config", () => ({
    BASE_URL: "https://test.com"
}));

describe("Home", () => {
    it("should render initially", () => {
        const spy = jest.spyOn(microsoftTeams, "initialize")
            .mockImplementation(() => undefined);
        const { container: { firstChild: component } } = render(<Home />);
        expect(component).toMatchSnapshot();
        spy.mockImplementation(callback => callback());
    });

    it("should initialize Microsoft Teams", () => {
        const spy = jest.spyOn(microsoftTeams.appInitialization, "notifySuccess");

        const { container: { firstChild: component } } = render(<Home />);
        expect(component).toMatchSnapshot();
        expect(spy).toBeCalledWith();
    });

    it("should set the validity state on input change", () => {
        const spy = jest.spyOn(microsoftTeams.settings, "setValidityState");
        const { container } = render(<Home />);

        const input = container.querySelector("input");

        expect(input.value).toBe("");
        fireEvent.change(input, { target: { value: "Good Day" } });
        expect(input.value).toBe("Good Day");
        expect(spy).toBeCalledWith(true);
    });

    it("should remove successfully", () => {
        const event = {
            notifySuccess: jest.fn(),
            notifyFailure: jest.fn()
        };
        const spy = jest.spyOn(microsoftTeams.settings, "registerOnRemoveHandler")
            .mockImplementation(callback => callback(event));

        render(<Home />);

        expect(event.notifySuccess).toBeCalledWith();
        expect(event.notifyFailure).not.toBeCalled();
        spy.mockImplementation();
    });

    it("should save successfully", () => {
        const event = {
            notifySuccess: jest.fn(),
            notifyFailure: jest.fn()
        };
        let saveCallback;
        const spySaveHandler = jest.spyOn(microsoftTeams.settings, "registerOnSaveHandler")
            .mockImplementation(callback => {
                saveCallback = callback;
            });
        const spySetSettings = jest.spyOn(microsoftTeams.settings, "setSettings");

        const { container } = render(<Home />);
        const input = container.querySelector("input");

        fireEvent.change(input, { target: { value: "Good Day" } });
        expect(input.value).toBe("Good Day");
        saveCallback(event);

        expect(event.notifySuccess).toBeCalledWith();
        expect(event.notifyFailure).not.toBeCalled();
        expect(spySetSettings).toBeCalledWith({
            configName: "Good Day",
            contentUrl: "https://test.com/?name={loginHint}&tenant={tid}&group={groupId}&theme={theme}",
            entityId: "Good Day"
        });

        spySaveHandler.mockImplementation();
    });
});
