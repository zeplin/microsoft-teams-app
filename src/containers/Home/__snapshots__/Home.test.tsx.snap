// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should initialize Microsoft Teams 1`] = `
<div
  class="ui-provider"
  data-uses-unhanded-props="true"
  dir="ltr"
>
  <div
    class="ui-flex"
    data-uses-unhanded-props="true"
  >
    <div
      aria-disabled="false"
      class="ui-box ui-input"
      data-aa-class="Input"
      data-uses-unhanded-props="true"
    >
      <input
        class="ui-box ui-input__input"
        data-uses-unhanded-props="true"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`Home should render initially 1`] = `
<div
  class="ui-provider"
  data-uses-unhanded-props="true"
  dir="ltr"
>
  <div
    class="ui-flex"
    data-uses-unhanded-props="true"
  >
    <div
      aria-disabled="true"
      class="ui-box ui-input"
      data-aa-class="Input"
      data-uses-unhanded-props="true"
    >
      <input
        class="ui-box ui-input__input"
        data-uses-unhanded-props="true"
        disabled=""
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
